internal/prometheusbp/spectest/spec_test.go

```diff
@@ -0,0 +1,340 @@
+		want           map[string]struct{}
+	}{
+		{
+			name:           "latency_seconds labels",
+			metricName:     "test_latency_seconds",
+			prefix:         "test",
+			clientOrServer: server,
+			want:           map[string]struct{}{"test_method": {}, "test_service": {}, "test_success": {}},
+		},
+		{
+			name:           "requests_total labels",
+			metricName:     "test_requests_total",
+			prefix:         "test",
+			clientOrServer: client,
+			want: map[string]struct{}{
+				"test_method":                {},
+				"test_service":               {},
+				"test_success":               {},
+				"test_baseplate_status":      {},
+				"test_baseplate_status_code": {},
+				"test_exception_type":        {},
+				"test_slug":                  {},
+			},
+		},
+		{
+			name:           "active_requests labels",
+			metricName:     "test_active_requests",
+			prefix:         "test",
+			clientOrServer: server,
+			want:           map[string]struct{}{"test_method": {}, "test_service": {}},
+		},
+		{
+			name:           "none",
+			prefix:         "test",
+			clientOrServer: server,
+			want:           map[string]struct{}{},
+		},
+	}
+	for _, tt := range testCases {
+		t.Run(tt.name, func(t *testing.T) {
+			if got, want := buildLables(tt.metricName, tt.prefix, tt.clientOrServer), tt.want; !reflect.DeepEqual(got, want) {
+				t.Fatalf("got %v, want %v", got, want)
+			}
+		})
+	}
+}
+
+func TestValidateSpec(t *testing.T) {
+	var (
+		testLabels = []string{
+			"thrift_method",
+			"thrift_service",
+			"thrift_success",
+			"thrift_baseplate_status",
+		}
+
+		testMetric = promauto.NewCounterVec(prometheus.CounterOpts{
+			Name: "thrift_client_latency_seconds",
+			Help: "Test help message",
+		}, testLabels)
+	)
+
+	labels := prometheus.Labels{
+		"thrift_method":           "foo",
+		"thrift_service":          "foo",
+		"thrift_success":          "foo",
+		"thrift_baseplate_status": "foo",
+	}
+	testMetric.With(labels).Inc()
+
+	testCases := []struct {
+		name           string
+		metric         prometheus.Collector
+		prefix         string
+		clientOrServer string
+		wantCount      int
+		wantErrs       []error
+	}{
+		{
+			name:           "not found server",
+			metric:         testMetric,
+			prefix:         "thrift",
+			clientOrServer: "server",
+			wantErrs:       []error{errNotFound},
+		},
+		{
+			name:           "multi errs client",
+			metric:         testMetric,
+			prefix:         "thrift",
+			clientOrServer: "client",
+			wantErrs:       []error{errPrometheusLint, errDiffLabels, errNotFound},
+		},
+	}
+
+	for _, tt := range testCases {
+		t.Run(tt.name, func(t *testing.T) {
+			gotErr := validateSpec(tt.prefix, tt.clientOrServer)
+			checkErrors(t, gotErr, tt.wantErrs)
+		})
+	}
+}
+
+func checkErrors(tb testing.TB, gotErr error, wantErrs []error) {
+	tb.Helper()
+
+	switch len(wantErrs) {
+	case 0:
+		if gotErr != nil {
+			tb.Errorf("expected nil err, got: %v", gotErr)
+		}
+	case 1:
+		// no op if only 1 error is expected
```

Comment:
you probably want to verify that `gotErr` is not a batch here, or wait for https://github.com/reddit/baseplate.go/pull/448 and cherry pick that into prometheus branch after it's merged.